### PR TITLE
Replace System.currentTimeMillis() by System.nanoTime()

### PR DESCRIPTION
**Pull request by wilderrodrigues**
- System.nanoTime() is the best way to measure elapsed time in Java.
   - It gives a resolution on the order of microseconds
   - The System.currentTimeMillis() is used when calculating absolut time.

Add unit tests to cover negative cases
   - Cover when the profile is not started/stopped

Test Environment
  - Management Server: CentOS 7.1
  - KVM

Tests successfully executed:

> nosetests --with-marvin --marvin-config=/data/shared/marvin/mct-zone2-kvm2-ISOLATED.cfg -s -a tags=advanced,required_hardware=false \
>     smoke/test_vm_life_cycle.py smoke/test_routers.py smoke/test_vpc_vpn.py smoke/test_reset_vm_on_reboot.py smoke/test_service_offerings.py \
>     component/test_vpc_offerings.py smoke/test_privategw_acl.py smoke/test_network_acl.py

### CODE REVIEW

File: utils/src/com/cloud/utils/Profiler.java [hunk 1]
```diff
@@ -23,24 +23,20 @@
...
     private Long stopTickInMs;
 
-    public Profiler() {
-        startTickInMs = null;
-        stopTickInMs = null;
-    }
-
     public long start() {
-        startTickInMs = System.currentTimeMillis();
+        startTickInMs = System.nanoTime();
```
Comment (DaanHoogland) [inline utils/src/com/cloud/utils/Profiler.java]: I'd use a new name for the var, like startTickInNanos (and of course stopTickInNanos)

File: utils/src/com/cloud/utils/Profiler.java [hunk 2]
```diff
@@ -23,24 +23,20 @@
...
     public long stop() {
-        stopTickInMs = System.currentTimeMillis();
+        stopTickInMs = System.nanoTime();
         return stopTickInMs.longValue();
     }
 
     public long getDuration() {
-        if (startTickInMs != null && stopTickInMs != null)
+        if (startTickInMs != null && stopTickInMs != null) {
             return stopTickInMs.longValue() - startTickInMs.longValue();
```
Comment (DaanHoogland) [inline utils/src/com/cloud/utils/Profiler.java]: hasn't the semantics changed as it no longer return milliseconds but now micro-(or even nano-) seconds? (this goes for stop and start as well btw. Did you check the callers?

File: utils/src/com/cloud/utils/Profiler.java (hunk 1 above)
Comment (wilderrodrigues) [inline utils/src/com/cloud/utils/Profiler.java]: The nanoTime() is given in Microseconds. So, the MS still fits in.

Cheers,
Wilder

File: utils/src/com/cloud/utils/Profiler.java (hunk 2 above)
Comment (wilderrodrigues) [inline utils/src/com/cloud/utils/Profiler.java]: I did... and it doesn't change. Only the resolution does,

Quick googled that one here for you: https://blogs.oracle.com/dholmes/entry/inside_the_hotspot_vm_clocks

Implementation is still the same:

image

http://stackoverflow.com/questions/351565/system-currenttimemillis-vs-system-nanotime

Cheers,
Wilder

Comment (DaanHoogland) [inline utils/src/com/cloud/utils/Profiler.java]: Yes, and so does the semantics of the return of starts() stop() and getDuration().
I checked; the first two are never cought in a variable but the last one is used in output and to check against settings. There the semantics really changes. It now returns higher numbers then axpected.

**Comment by wilderrodrigues** [comment]
Before we continue in a face-to-face chat, which would be better, just few point to take into account:

* The Profiler class is used as a profiler, to just measure how long a code took to be executed; used along with sleep calls - like a stop watch; and, still, as a monitor - there are lock implementations based on it. It makes the Profile class a profile, a stop-watch and a monitor.

* From a profiler perspective, do a Star, wait for 1 second, then a Stop and check how long it took to compute doesn't seem right to me.

The comments above are only related to the fact that the Profiler class is being exploited inside ACS.

Concerning the getDuration(). I agree that once it is taken into account to calculate milliseconds, it will be flaw.

In that case, it's too bad I trusted the existing test - which has a glitch in the assertion.

My suggestion is to refactor the getDuration(), renaming it into getDurationInMillis() and create a getDuration() which returns the nanotime - microseconds. The test has to be fixed as well.
